### DISCUSSION TITLE
The portable `yad` almost works, but only `yad` command works, and some failures (...)

### DISCUSSION THREAD

**Discussion by gusbemacbe** [flatimage/flatimage#5] [Q&A]
Hello!

I managed to build a portable container on Red Hat Linux 8.7. The complete process took more than 3 hours because the command line `Excepting share-mime-info` caused a stuckness in the terminal and a slowdown in the host OS. 

I symlinked `yad.arts` to `yad`, `yad-icon-browser` and `yad-tools`:

* `yad --help` works and shows the correct details;
* `yad-icon-browser --help` shows the incorrect details;
* `yad-tools --help` has the same problem as `yad-icon-browser`;
* `yad-icon-browser` cannot open Yad's GUI. 

1. I would like to resize from 10G to 600 MB. Is there how?
2. If it works on Red Hat Linux, will it also work on any old and new distributions versions, as Arch Linux?

**Comment by ruanformigoni** [comment]
Hey, thanks for the feedback! 

The symlinks did not work because arts executes the command you set with 'arts-cmd', symlinks is a WIP, for now you can create a script called `yad-icon-browser` that does:

```sh
#!/bin/bash

HERE=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )

"$HERE"/yad.arts arts-exec "$(basename "$0") $*"
```

Put his script in the same directory as `yad.arts` and use it as you would a symlink with `./yad-icon-browser`. For the other tools, you can just copy this script with the desired tool name `cp yad-icon-browser yad-tools`.

---

The image apparent size is 10G, you can see the actual size with `du -sh ./yad.arts`. To resize it you can use `arts-compress`, this will make the image immutable, so make sure to install everything you need in it; or you can try to use `./yad.arts arts-resize 600MB`.

---

I'll check the delay with the `share-mime-info` issue, thanks for bringing this to my attention.

**Comment by gusbemacbe** [comment]
As Alpine does not have `xcolor` in their official repositories, I used Arch. But it failed:

![image](https://user-images.githubusercontent.com/2257024/233515512-8ff7316c-c4eb-453c-bedc-d58264f9e91f.png)

**Reply by ruanformigoni** [reply]
Arch requires the `/var/lib/pacman` dir to exist, for that you need to run `sudo mkdir -p /var/lib/pacman`. If you have no superuser rights, you can try to use focal, which does not have this limitation.

**Reply by gusbemacbe** [reply]
Ah, it worked now. I have superuser rights as I am using RHEL 8.7 image on VirtualBox on my personal computer.

**Reply by ruanformigoni** [reply]
> If it works on Red Hat Linux, will it also work on any old and new distributions versions, as Arch Linux?

Forgot to answer this question, yes, the images should work on new and old distributions. If it has a kernel that is too old, it might have issues however.

**Reply by gusbemacbe** [reply]
The `xcolor` still failed. I ran the following commands:

```shell
export ARTS_RW="0.5.1"
./arch.arts arts-resize 3G
./arch.arts arts-root pacman -S xcolor --noconfirm
mv arch.arts xcolor.arts
./xcolor.arts arts-cmd xcolor
./xcolor.arts arts-exec xcolor
./xcolor.arts -s clipboard
```

See the screenshot:

![image](https://user-images.githubusercontent.com/2257024/233517218-4b6e129d-c9d5-4842-ac29-cf21a72a2d92.png)

Did I do something wrong?

**Reply by ruanformigoni** [reply]
If you execute `echo "$XDG_SESSION_TYPE"` you can see that rhel runs `wayland`, not `X11`.  This is probably why this error happens, you can try to use a wayland color picker, as [wl-color-picker](https://github.com/jgmdev/wl-color-picker).

Here's how:

```
./arch.arts arts-root pacman -S grim imagemagick zenity wl-clipboard wget
./arch.arts arts-root wget -O/usr/bin/wcolor https://raw.githubusercontent.com/jgmdev/wl-color-picker/main/wl-color-picker.sh
./arch.arts arts-exec chmod +x /usr/bin/wcolor
./arch.arts arts-exec wcolor
```

~~However the 'picking' functionality seems to have an issue with the compositor for this software + gnome.~~

**Reply by gusbemacbe** [reply]
Ah, I have to test on X11 because X11 is the default desktop manager of my company's distribution. To be agnostic to Wayland or X11, I think Farge is better.

**Reply by ruanformigoni** [reply]
For reference, got it to work, instead of clicking on `select`, click on here:

![screenshot-171](https://user-images.githubusercontent.com/18180693/233520984-deec4e81-68eb-4e39-b2fc-b4e62d94678a.png)

**Reply by ruanformigoni** [reply]
Updated the `arch.arts` release. Now you should be able to run it in systems without the `/var/lib/pacman` dir (the warning will still show).

**Reply by gusbemacbe** [reply]
Can I remove the  `.yay` folder from the `.cache` folder and running `paccache` before running `arts-compress`?

**Reply by ruanformigoni** [reply]
Yes, but the yay cache is not copied to inside the container if it is in your `home` directory.

**Reply by gusbemacbe** [reply]
I use `arch.arts`. The `xcolor.arts` worked perfectly on Red Hat Linux 8.7, but it does not work here on the latest version of Endeavour OS, derived from pure Arch Linux. Here are the errors:

![image](https://user-images.githubusercontent.com/2257024/233528485-1a454ee3-ab46-4952-bfd8-d7511733ce49.png)

**Reply by ruanformigoni** [reply]
Ok, I'll try to reproduce the error here in a VM.

**Reply by ruanformigoni** [reply]
Does it work if it is uncompressed? Also, could you tell what is the glibc version in your system, with `ldd --version`?

**Reply by gusbemacbe** [reply]
> Does it work if it is uncompressed?

I am unable to mount the `arts` as it shows the same errors. 

> Also, could you tell what is the glibc version in your system, with `ldd --version`?

```
ldd (GNU libc) 2.37
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

**Reply by ruanformigoni** [reply]
<img src="https://user-images.githubusercontent.com/18180693/233538514-feb47cf9-3924-4ddb-99e4-1bc5942dfd3b.png" width=600px>

Worked fine for me in a VM. Could you test the `ext2rd` binary available [here](https://gitlab.com/formigoni/arts/-/jobs/4152494818/artifacts/browse/bin/), I think it is what might be what is failing. Just run `chmod +x ext2rd` and `./ext2rd` to check if it fails with the same error messages.

**Reply by gusbemacbe** [reply]
It still shows the error messages:

![image](https://user-images.githubusercontent.com/2257024/233539471-3eb87fef-b379-4d0c-ae21-037eb91106b2.png)

**Reply by ruanformigoni** [reply]
Thanks a ton for your feedback :smile: I have updated the ext2rd used in arts [here](https://github.com/ruanformigoni/extfstools/releases/tag/58d896b), could you test to see if it works now? If it does I'll update the `arts` packages to use this one instead.

**Reply by gusbemacbe** [reply]
It works now.

**Reply by ruanformigoni** [reply]
New version with the fixed ext2rd binary released.

**Reply by gusbemacbe** [reply]
Strange, the latest version of `ext2rd` worked perfectly here, but the `alpine.arts` and `arch.arts` still did not work:

![image](https://user-images.githubusercontent.com/2257024/233546736-224a4719-2cbf-4f24-ab52-fa79e7071c27.png)

**Reply by ruanformigoni** [reply]
Could you try to remove the `arts` cache? It's `rm -rf /tmp/arts`.

**Reply by gusbemacbe** [reply]
It works now. It was the `/tmp/`

**Reply by gusbemacbe** [reply]
Strange, `arts-resize` does not resize, the `arch.arts` is still 1 GB after having run the `arts-resize 3G`.

**Reply by ruanformigoni** [reply]
Sorry about that, could you post the output of the command?

**Reply by gusbemacbe** [reply]
When I ran the `./arch.arts arts-resize 3G`, there was no error, it just did not give effect.

**Reply by ruanformigoni** [reply]
I was probably `resize2fs` that failed, you can get more output with `ARTS_DEBUG=1 ./arch.arts arts-resize 3G`

**Reply by gusbemacbe** [reply]
```
[*] ARTS_RO          : 
[*] ARTS_RW          : 1
[*] ARTS_STREAM      : /dev/stdout
[*] ARTS_ROOT        : 
[*] ARTS_NORM        : 1
[*] ARTS_DEBUG       : 1
[*] ARTS_NDEBUG      : 
[*] ARTS_BIN         : /tmp/arts
[*] ARTS_OFFSET      : 18181128
[*] ARTS_MOUNT       : /tmp/tmpdir.DFnoIG/mount
[*] ARTS_TEMP        : /tmp/tmpdir.DFnoIG
[*] ARTS_FILE        : /home/benegus/.local/arch.arts
[*] $*               : arts-resize 3G
e2fsck 1.46.6-rc1 (12-Sep-2022)
Pass 1: Checking inodes, blocks, and sizes
Pass 2: Checking directory structure
Pass 3: Checking directory connectivity
Pass 4: Checking reference counts
Pass 5: Checking group summary information
/home/benegus/.local/arch.arts: 38309/67072 files (0.4% non-contiguous), 965686/1068032 blocks
Error relocating /lib//libfuse.so.2: __snprintf_chk: symbol not found
Error relocating /lib//libfuse.so.2: __realpath_chk: symbol not found
Error relocating /lib//libfuse.so.2: __memcpy_chk: symbol not found
Error relocating /lib//libfuse.so.2: fcntl64: symbol not found
Error relocating /lib//libfuse.so.2: __fprintf_chk: symbol not found
Error relocating /lib//libfuse.so.2: __sprintf_chk: symbol not found
fusermount: entry for /tmp/tmpdir.DFnoIG/mount not found in /etc/mtab
```

**Reply by ruanformigoni** [reply]
Thanks for the output :+1: , it seems to be the same error as before, I'll make the change that I did for ext2rd for the other used libraries and integrate into a new release.

**Reply by ruanformigoni** [reply]
Meanwhile, you can try to build it in a VM and copy it in the main host afterwards.

**Reply by ruanformigoni** [reply]
I've updated the other binaries with the fix, except the compression one.

**Reply by gusbemacbe** [reply]
I have just tested. It worked perfectly. The download and the installation are quicker on the native OS than on the VM. It is a pity that I cannot compress, as you still did not fix the compression binary. 

Observe that after running the command `./arch.arts arts-exec xcolor -s clipboard` and after selecting a colour, the command was not stopped or terminated, it was still stuck. I had to copy the random word from another window to stop the command.

**Reply by ruanformigoni** [reply]
Glad to hear it works! Currently working on the compression binaries. I as able to reproduce the xcolor issue, will investigate further.

**Reply by ruanformigoni** [reply]
Hello, compression binaries updated in the latest release.

**Comment by gusbemacbe** [comment]
Hello @ruanformigoni !

Has the extension `arts` been renamed to `flatimage`?

**Reply by ruanformigoni** [reply]
Hi there! Yes, I've renamed the project :grin:
